repo2docker/buildpacks/base.py

```diff
@@ -349,7 +350,21 @@ def _filter_tar(tar):
             src_path = os.path.join(os.path.dirname(__file__), *src_parts)
             tar.add(src_path, src, filter=_filter_tar)
 
-        tar.add('.', 'src/', filter=_filter_tar)
+        _exclude_tar = None
+        if os.path.exists(".gitignore"):
+            gitignore_fh = open(".gitignore")
+            ignorespec = pathspec.PathSpec.from_lines('gitignore', gitignore_fh)
+
+            def _exclude_tar(filepath):
+                # Conditionally exclude files based on the pathspecs mentioned
+                # in the `.gitignore` file.
+                # Note that, the behaviour of this function is not **exactly**
+                # same to the way `git` excludes files based on `.gitignore`.
+                #
+                # https://github.com/cpburnz/python-path-specification/issues/19
+                return ignorespec.match_file(filepath)
```

Comment:
It concerns me a bit that this isn't exact. Is there any possibility that we could end up excluding files that shouldn't be excluded, or does the bug you link exclusively mean that we will include some files we maybe shouldn't?

Reply:
@minrk : `pathspecs` was the only pure python implementation I could find. Also, I think it does implement the `pathspecs` correctly, but the difference comes from how `git` traverses the directory structure when matching them with the `pathspecs` (eg: if a directory is ignored, then it doesnot traverse it, on the other hand in this use case, `tarfile` does traverse the whole directory and tries to match with all its children too).

In the issue I reported, and in most of the use cases, this will only mean we will _forget_ to exclude some files which were meant to be excluded. 

The only use case when a file that is not meant to be excluded is excluded, is when `.gitignore` specifically ignores a particular file, but the user had force-pushed the file to the repository.

But in the both these use cases, the `.gitignore` should ideally be modified to include the exceptions. For example, when you force-push say a `*.csv` file which was ignored by `.gitignore`, then you are expected to also remember to add a `!<path_to_file>` to `.gitignore`.